lib/model/channel.dart

```diff
@@ -368,6 +405,49 @@ class ChannelStoreImpl extends HasUserStore with ChannelStore {
   @override
   final Map<int, ChannelFolder> channelFolders;
 
+  /// Maps indexed by channel IDs, of the known latest message IDs in each topic.
+  ///
+  /// For example: `_latestMessageIdsByChannelTopic[channel.streamId][topic] = maxId`
+  ///
+  /// Occasionally, the latest message ID of a topic will refer to a message
+  /// that doesn't exist or is no longer in the topic.
+  /// This happens when the topic's latest message is deleted or moved
+  /// and we don't have enough information to replace it accurately.
+  /// (We don't keep a snapshot of all messages.)
+  // TODO(#2004): handle more cases where this can change
+  final Map<int, TopicKeyedMap<int>> _latestMessageIdsByChannelTopic = {};
+
+  @override
+  Future<void> fetchChannelTopics(int channelId) async {
+    if (_latestMessageIdsByChannelTopic[channelId] != null) return;
+
+    final result = await _apiGetChannelTopics(connection, channelId: channelId,
+      allowEmptyTopicName: true);
+    (_latestMessageIdsByChannelTopic[channelId] = makeTopicKeyedMap())
+      .addEntries(result.topics.map((entry) => MapEntry(entry.name, entry.maxId)));
```

Comment:
If this method gets called while another call is already in the middle of a request, it'll make a fresh request. Then whichever request finishes later will clobber the results of the one that finished first.

Is that a situation that's likely to happen in practice? What would the consequences be?

If we want to handle that situation smoothly, see the logic in GlobalStore.perAccount around `_perAccountStoresLoading`.